### ISSUE TITLE
Iterating over an object with for of causes infinite loop when target is ES5

### ISSUE THREAD

**Issue by alexanderzobnin** [microsoft/TypeScript#30423]
**TypeScript Version:**  3.0.3
tsconfig:
```
{
  "compilerOptions": {
    "moduleResolution": "node",
    "target": "es5",
...
```

**Search Terms:** __values, for of loop, for of loop iterable, 

**Code**
We're doing something like this:
```ts
for (const query of targets) {
  if (!query.refId) {
    query.refId = this.getNextQueryLetter();
  }
}
```
`targets` is an array of objects and comes from outside.

**Expected behavior:**
If `targets` is not an Array, but an Object it shouldn't be iterable. In the ES6 env this throws an exception `TypeError: targets is not iterable`.

**Actual behavior:**
There's no exception and code goes into an infinite loop.

The reason is fallback `for of` implementation (valuesHelper):
https://github.com/Microsoft/TypeScript/blob/master/src/compiler/factory.ts#L3613
```js
var __values = (this && this.__values) || function (o) {
  var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
  if (m) return m.call(o);
  return {
      next: function () {
          if (o && i >= o.length) o = void 0;
          return { value: o && o[i++], done: !o };
      }
  };
};
```
In case of object type `return { value: o && o[i++], done: !o };` will be invoked infinitely. This `if (o && i >= o.length) o = void 0;` will never be true since `o.length` is `undefined`.

**Related Issues:** 
This issue may be related as well #20394

**Comment by RyanCavanaugh** [comment]
What is `targets`? You said `targets is an array of objects and comes from outside.` but also `If targets is not an Array, but an Object it shouldn't be iterable. `

How do we reproduce this behavior?

**Comment by alexanderzobnin** [comment]
Normally, `targets` is Array. But there're situations where `targets` is an Object. In order to reproduce this just try to iterate over an Object:
```ts
const targets = { foo: "valueFoo", bar: "valueBar" };
for (const query of targets) {
  // do something
  console.log(query);
}
```

**Comment by RyanCavanaugh** [comment]
@rbuckton proposal here is that we harden the iteration helper to behave correctly if given an incorrectly-typed value. I think we can just add `if (o.length === void) throw o` above the `return` ?

**Comment by rbuckton** [comment]
Fixed by #32468
